16-H0ngJu

H0ngJu opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[뱀과 사다리 게임](https://www.acmicpc.net/problem/16928)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
40M

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

이번 문제는 금방 로직을 짰습니다 허허
<br/>

**_문제 요약_**
뱀-사다리 게임에서 주사위를 조작해서 100까지 도달하자@! 이때 주사위를 굴리는 최소 횟수는?
<br/><br/>

<details>
    <summary>알고리즘 분류</summary>

음 문제 보고 바로 오 이거 BFS다~로 생각해서 덱들고 바로 접근했습니다 🔥

그냥 흔한 BFS 문제와 비슷해서 로직을 짜는 데 크게 주의해야할 부분은 없었습니다


다만, 제 로직에서는
사다리가 있는지, 뱀이 있는지, 그냥 일반 칸인지 구분하기 위해
`check`이라는 flag를 따로 두었습니다.
<br/>
예를 들어,

현재 칸은 40, 주사위는 3이고, 43번째 칸이 `사다리`칸이라면, 
`뱀`, `일반`칸 로직은 확인할 필요가 없으므로
check = 1로 바꾸어 주고, 다음 주사위를 굴리면 됩니다.

다음 주사위를 굴리게되면 주사위가 4가 되므로, 44번째 칸에 대한 로직을 진행하면 됩니다


</details>


<br/><br/>

**_최종 수도 코드_**
1. q가 빌 때까지 아래를 반복
    1. cur(현재 위치)가 100이면 종료.
    2. 사다리, 뱀, 일반 칸인지 확인
    3. 각 케이스에 대하여 방문하지 않았으면 q에 삽입
    4. check = 1로 두어 불필요한 반복문 실행 줄임

<br/><br/>
```python
import sys
from collections import deque

def input(): return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
ladder = [list(map(int, input().split())) for _ in range(N)]
snake = [list(map(int, input().split())) for _ in range(M)]
game = [i for i in range(1,101)]
visited = [0 for _ in range(100)]
q = deque()

q.append((1,0))
visited[0] = 1

while q:
    cur, dice = q.popleft()
    if cur == 100:
        print(dice)
        break
    for i in range(1, 7):
        check = 0
        
        if cur + i > 100: continue

        # 사다리 갈 수 있는지 검사
        for l, n in ladder:
            if cur + i == l:
                if visited[n-1] == 0:
                    q.append((n, dice+1))
                    visited[n-1] = 1
                check = 1
                break

        if check: continue
        # 뱀 있는지 검사
        for s, n in snake:
            if cur + i == s:
                if visited[n-1] == 0:
                    q.append((n, dice+1))
                    visited[n-1] = 1
                check = 1
                break

        if check: continue 
        # 그 외 
        if visited[cur + i-1] == 0:
            q.append((cur + i, dice + 1))
            visited[cur + i-1] = 1
```

## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

9kyo-hwang reviewed (approved):
1. 뱀 또는 사다리는 "반드시" 타야합니다. 또한 각 칸 별로 뱀 또는 사다리가 최대 1개 존재합니다.
따라서 뱀과 사다리에 대한 정보를 딕셔너리로 저장해놓고, 현재 칸에 뱀 또는 사다리가 있다면 그거를 타고 아래 로직은 무시하도록 해버리면 코드를 깔끔하게 만들 수 있습니다.

2. 최종적으로 우리가 원하는 건 주사위를 굴리는 횟수입니다. 
이는 일종의 최단 경로 문제로 볼 수 있는데, visited 리스트와 deque에 주사위 굴린 횟수를 별도로 저장하지 않고 distances란 리스트 하나로 모두 해결 가능합니다.

```py
from collections import deque
input = open(0).readline

N, M = map(int, input().split())
distances = [float('inf')] * 101  # 각 칸을 오기까지 굴린 주사위 최소 횟수

ladder_and_snake = {}  # 뱀과 사다리는 각 칸 별 최대 1개
for _ in range(N + M):
    a, b = map(int, input().split())
    ladder_and_snake[a] = b  # 딕셔너리 형태로 한 번에 저장

q = deque([1])  # 최초 시작점
distances[1] = 0  # 시작점에서의 주사위를 굴린 횟수는 0

while q:
    u = q.popleft()  # 현재 칸 위치
    if u == 100:  # 도착지면 stop
        break
    
    if u in ladder_and_snake:  # 우선 현재 칸에 뱀 또는 사다리가 있는 지 확인
        v = ladder_and_snake[u]
        if distances[u] < distances[v]:  # 목적지 칸까지 가는 데 굴린 주사위 횟수가 더 많은 경우에만
            distances[v] = distances[u]  # 뱀, 사다리는 주사위를 굴리지 않으므로 값 그대로 들어감
            q.append(v)
        continue  # 있다면 반드시 타야하므로 아래 주사위를 굴려보는 로직에 접근하지 않음

    for dice in range(1, 7):  # 뱀, 사다리가 아니면 주사위를 굴림
        v = u + dice  # 다음 칸은 현재 칸 + 주사위 눈
        if v <= 100 and distances[u] + 1 < distances[v]:  # 100을 넘지 않으면서 굴린 횟수가 더 적은 경우에만
            distances[v] = distances[u] + 1  # 주사위를 굴렸으므로 + 1
            q.append(v)
            
print(distances[100])  # while문이 종료되면 목적지 칸에 저장된 주사위 횟수 출력.
```

재밌는 문제였네요 :D

SeongHoonC commented on H0ngJu/BFS/뱀과 사다리 게임.py:
```diff
@@ -0,0 +1,49 @@
+def input(): return sys.stdin.readline().rstrip()
+
+N, M = map(int, input().split())
+ladder = [list(map(int, input().split())) for _ in range(N)]
+snake = [list(map(int, input().split())) for _ in range(M)]
```
저도 하나로 했습니둥

SeongHoonC commented on H0ngJu/BFS/뱀과 사다리 게임.py:
```diff
@@ -0,0 +1,49 @@
+        
+        if cur + i > 100: continue
+
+        # 사다리 갈 수 있는지 검사
+        for l, n in ladder:
+            if cur + i == l:
+                if visited[n-1] == 0:
+                    q.append((n, dice+1))
+                    visited[n-1] = 1
+                check = 1
+                break
+
+        if check: continue
+        # 뱀 있는지 검사
+        for s, n in snake:
+            if cur + i == s:
+                if visited[n-1] == 0:
+                    q.append((n, dice+1))
+                    visited[n-1] = 1
+                check = 1
+                break
+
```
하나로 합칠 수 있겠네용

H0ngJu replied:
그러네요 이제 check도 필요 없네요 .. 🔨🔨

SeongHoonC reviewed (approved):
구현은 안어려웠는데 bfs 임을 떠울리는게 젤 어려웠네요.. 숨바꼭질 문제랑 비슷한 것 같아요!!

<img width="663" alt="image" src="https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/108349655/549a8323-318e-452f-bf46-f85c5c329c21">

```kotlin
import java.io.BufferedReader
import java.io.InputStreamReader

const val MAX = 987_654_321
fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.getInputs()
    val arr = Array(101) { MAX }
    val q = ArrayDeque<Int>()

    val letterOrSnake = mutableMapOf<Int, Int>()
    repeat(n) {
        val (x, y) = br.getInputs()
        letterOrSnake[x] = y
    }
    repeat(m) {
        val (u, v) = br.getInputs()
        letterOrSnake[u] = v
    }

    arr[1] = 0
    q.add(1)

    while (q.isNotEmpty()) {
        val now = q.removeFirst()
        val nextOrNull = letterOrSnake[now]
        if (nextOrNull != null) {
            if (arr[nextOrNull] > arr[now]) {
                arr[nextOrNull] = arr[now]
                q.add(nextOrNull)
            }
            continue
        }
        for (i in 1..6) {
            val next = now + i
            if (next > 100) {
                continue
            }
            if (arr[next] > arr[now] + 1) {
                arr[next] = arr[now] + 1
                q.add(next)
            }
        }
    }
    println(arr[100])
}

private fun BufferedReader.getInputs() = readLine().split(" ").map { it.toInt() }
```

tgyuuAn commented:
사다리가 ladder 얐군요....?
ㅋㅋㅋㅋㅋㅋ
@SeongHoonC ㅋㅋㅋㅋㅌㅋㅋㅋㅋㅋㅋㅋㅋ저도 변수명 틀림욬ㅋㅋㅋㅋ

H0ngJu commented:
> 1. 뱀 또는 사다리는 "반드시" 타야합니다. 또한 각 칸 별로 뱀 또는 사다리가 최대 1개 존재합니다.
>    따라서 뱀과 사다리에 대한 정보를 딕셔너리로 저장해놓고, 현재 칸에 뱀 또는 사다리가 있다면 그거를 타고 아래 로직은 무시하도록 해버리면 코드를 깔끔하게 만들 수 있습니다.
> 2. 최종적으로 우리가 원하는 건 주사위를 굴리는 횟수입니다.
>    이는 일종의 최단 경로 문제로 볼 수 있는데, visited 리스트와 deque에 주사위 굴린 횟수를 별도로 저장하지 않고 distances란 리스트 하나로 모두 해결 가능합니다.
> 
> ```python
> from collections import deque
> input = open(0).readline
> 
> N, M = map(int, input().split())
> distances = [float('inf')] * 101  # 각 칸을 오기까지 굴린 주사위 최소 횟수
> 
> ladder_and_snake = {}  # 뱀과 사다리는 각 칸 별 최대 1개
> for _ in range(N + M):
>     a, b = map(int, input().split())
>     ladder_and_snake[a] = b  # 딕셔너리 형태로 한 번에 저장
> 
> q = deque([1])  # 최초 시작점
> distances[1] = 0  # 시작점에서의 주사위를 굴린 횟수는 0
> 
> while q:
>     u = q.popleft()  # 현재 칸 위치
>     if u == 100:  # 도착지면 stop
>         break
>     
>     if u in ladder_and_snake:  # 우선 현재 칸에 뱀 또는 사다리가 있는 지 확인
>         v = ladder_and_snake[u]
>         if distances[u] < distances[v]:  # 목적지 칸까지 가는 데 굴린 주사위 횟수가 더 많은 경우에만
>             distances[v] = distances[u]  # 뱀, 사다리는 주사위를 굴리지 않으므로 값 그대로 들어감
>             q.append(v)
>         continue  # 있다면 반드시 타야하므로 아래 주사위를 굴려보는 로직에 접근하지 않음
> 
>     for dice in range(1, 7):  # 뱀, 사다리가 아니면 주사위를 굴림
>         v = u + dice  # 다음 칸은 현재 칸 + 주사위 눈
>         if v <= 100 and distances[u] + 1 < distances[v]:  # 100을 넘지 않으면서 굴린 횟수가 더 적은 경우에만
>             distances[v] = distances[u] + 1  # 주사위를 굴렸으므로 + 1
>             q.append(v)
>             
> print(distances[100])  # while문이 종료되면 목적지 칸에 저장된 주사위 횟수 출력.
> ```
> 
> 재밌는 문제였네요 :D

오홍 .. 딕셔너리로 저장하고 굳이 visited로 따로 방문 관리를 하지 않는다..!!
이해했습니다 !! 따로 coutn 안두는게 저번 @alstjr7437 토메이토 문제같군요

담에 써먹겠습니다 😎

alstjr7437 commented on H0ngJu/BFS/뱀과 사다리 게임.py:
```diff
@@ -0,0 +1,49 @@
+def input(): return sys.stdin.readline().rstrip()
+
+N, M = map(int, input().split())
+ladder = [list(map(int, input().split())) for _ in range(N)]
+snake = [list(map(int, input().split())) for _ in range(M)]
```
어짜피 받을때 start end는 같더라구요!!
사다리 먼저 받지만 사다리는 위로 올라가는데 
start -> end로 start < end가 되고

뱀은 아래로 내려가지만 input은 동일하게
start -> end로 start > end가 되는 다른점만 있을뿐 같더라구요!

alstjr7437 commented on H0ngJu/BFS/뱀과 사다리 게임.py:
```diff
@@ -0,0 +1,49 @@
+import sys
+from collections import deque
+
+def input(): return sys.stdin.readline().rstrip()
```
input 바꾸실때 이렇게 쓰시는군요!!
저는 
```python
input = sys.stdin.readline
```
으로 바꿔서 신기하네요!!

alstjr7437 commented:
홍주님 문제 이해하고 로직을 이해했는데 따로 리뷰를 늦게 달아서 그런 점도 있지만
홍주님 코드와 엄청 비슷하게 나오더라구요!!!

그래서 그냥 홍주님 코드에서 반복되는 부분 
1. snake, labber부분 
2. 사다리 뱀 검사 부분

등을 줄여서 해봤습니다!!!

```python
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())
unique = [list(map(int, input().split())) for _ in range(N+M)]
game = [i for i in range(1,101)]
visited = [0 for _ in range(100)]
q = deque()

q.append((1,0))
visited[0] = 1

while q:
    cur, dice = q.popleft()
    if cur == 100:
        print(dice)
        break
    for i in range(1, 7):
        check = 0
        
        if cur + i > 100: continue

        # 사다리 또는 뱀 검사
        for start, end in unique:
            if cur + i == start:
                if visited[end-1] == 0:
                    q.append((end, dice+1))
                    visited[end-1] = 1
                check = 1
                break

        if check: continue

        # 그 외 
        if visited[cur + i-1] == 0:
            q.append((cur + i, dice + 1))
            visited[cur + i-1] = 1
```

시간과 메모리는 막 엄청 큰차이가 있진 않네요!
<img width="921" alt="image" src="https://github.com/AlgoLeadMe/AlgoLeadMe-1/assets/94051599/98791cdf-81db-4a16-b37d-ecf8c4ab2222">
